src/vm/synch.cpp

```diff
@@ -668,8 +668,19 @@ bool CLRLifoSemaphore::WaitForSignal(DWORD timeoutMs)
 #endif // FEATURE_PAL
         }
 
+        if (!waitSuccessful)
+        {
+            // Unregister the waiter. The wait subsystem used above guarantees that a thread that wakes due to a timeout does
+            // not observe a signal to the object being waited upon.
+            Counts toSubtract;
+            ++toSubtract.waiterCount;
+            Counts countsBeforeUpdate = m_counts.ExchangeAdd(-toSubtract);
+            _ASSERTE(countsBeforeUpdate.waiterCount != (UINT16)0);
+            return false;
+        }
```

Comment:
This is essentially just a minor optimization for the non-successful/timeout case?  Presumably there's no harm in throughput to the successful case?

Reply:
It also doesn't attempt to acquire upon timeout since it wouldn't have observed a signal anyway, there was no change to perf